openff/bespokefit/executor/client.py

```diff
@@ -0,0 +1,292 @@
+    )
+
+    @property
+    def bespoke_force_field(self) -> Optional[ForceField]:
+        """The final bespoke force field if the bespoke fitting workflow is complete."""
+
+        if self.results is None or self.results.refit_force_field is None:
+            return None
+
+        return ForceField(
+            self.results.refit_force_field, allow_cosmetic_attributes=True
+        )
+
+    @property
+    def status(self) -> Status:
+        pending_stages = [stage for stage in self.stages if stage.status == "waiting"]
+
+        running_stages = [stage for stage in self.stages if stage.status == "running"]
+        assert len(running_stages) < 2
+
+        running_stage = None if len(running_stages) == 0 else running_stages[0]
+
+        complete_stages = [
+            stage
+            for stage in self.stages
+            if stage not in pending_stages and stage not in running_stages
+        ]
+
+        if (
+            running_stage is None
+            and len(complete_stages) == 0
+            and len(pending_stages) > 0
+        ):
+            return "waiting"
+
+        if any(stage.status == "errored" for stage in complete_stages):
+            return "errored"
+
+        if running_stage is not None or len(pending_stages) > 0:
+            return "running"
+
+        if all(stage.status == "success" for stage in complete_stages):
+            return "success"
+
+        raise NotImplementedError()
+
+    @property
+    def error(self) -> Optional[str]:
+        """The error that caused the fitting to fail if any"""
+
+        if self.status != "errored":
+            return None
+
+        message = next(
+            iter(stage.error for stage in self.stages if stage.status == "errored")
+        )
+        return "unknown error" if message is None else message
+
+    @classmethod
+    def from_response(cls: Type[_T], response: CoordinatorGETResponse) -> _T:
+        """Creates an instance of this object from the response from a bespoke
+        coordinator service."""
+
+        return cls(
+            smiles=response.smiles,
+            stages=[
+                BespokeExecutorStageOutput(
+                    type=stage.type, status=stage.status, error=stage.error
+                )
+                for stage in response.stages
+            ],
+            results=response.results,
+        )
+
+
+class BespokeFitClient:
+    """
+    A client interface which can be used to connect to a BespokeFit executor instance to submit and retrieve results.
+    """
+
+    def __init__(
+        self,
+        settings: Settings,
+        retries: int = 3,
+        backoff_factor: float = 0.5,
+    ):
+        self._session = requests.Session()
+        self._session.headers.update({"bespokefit-token": settings.BEFLOW_API_TOKEN})
+        retry = Retry(connect=retries, backoff_factor=backoff_factor)
+        adapter = HTTPAdapter(max_retries=retry)
+        self._session.mount("http://", adapter)
```

Comment:
@dotsdl if you have a moment - I know next to nothing about internet security, but this smells a little funny to me. If we're adding API tokens to bespokefit, is it safe to put them in headers over http, or should we be using https?

Reply:
Good point I would have expected that as well, I checked Alchemiscale and it turns out you can use something like [traefik](https://github.com/traefik/traefik) to automatically do this for you without having to implement it in the package. You can see that this is part of the deployment docker file in [alchemiscale](https://github.com/openforcefield/alchemiscale/blob/938690746692dbfa2d4b97c2423495ccb26a6b2b/docker/alchemiscale-server/docker-compose.yml#L172).

Reply:
@bennybp what are you using for handling TLS on your QCFractal deployments? We we `traefik` to handle this for `alchemiscale`, but curious what you've been using.

Reply:
Same here. All the instances (and other projects) that are on that server run behind traefik, which handles all the routing and certificates. QCFractal doesn't have direct support for SSL itself (at least I don't see it as on option for `waitress`)